src/app-layer-parser.c

```diff
@@ -1162,16 +1147,18 @@ int AppLayerParserSupportsFiles(uint8_t ipproto, AppProto alproto)
 DetectEngineState *AppLayerParserGetTxDetectState(uint8_t ipproto, AppProto alproto, void *tx)
 {
     SCEnter();
-    DetectEngineState *s = alp_ctx.ctxs[FlowGetProtoMapping(ipproto)][alproto].GetTxDetectState(tx);
+    AppLayerTxData *d = AppLayerParserGetTxData(ipproto, alproto, tx);
```

Comment:
I think this should be taken one step further: get rid of `AppLayerParserGetTxDetectState`. The detect engine should have direct access to `AppLayerTxData`.